skrifa/src/color/mod.rs

```diff
@@ -1,6 +1,437 @@
+    /// transformation matrix.
+    fn push_transform(&mut self, transform: Transform);
+    /// Restore the transformation matrix to the state before the previous
+    /// [`push_transform`](ColorPainter::push_transform) call.
+    fn pop_transform(&mut self);
+
+    /// Apply a clip path in the shape of glyph specified by `glyph_id`.
+    fn push_clip_glyph(&mut self, glyph_id: GlyphId);
+    /// Apply a clip rectangle specified by `clip_rect`.
+    fn push_clip_box(&mut self, clip_box: BoundingBox<f32>);
+    /// Restore the clip state to the state before a previous
+    /// [`push_clip_glyph`](ColorPainter::push_clip_glyph) or
+    /// [`push_clip_box`](ColorPainter::push_clip_box) call.
+    fn pop_clip(&mut self);
+
+    /// Fill the current clip area with the specified gradient fill.
+    fn fill(&mut self, fill_type: FillType);
+
+    /// Optionally implement this method: Draw an unscaled COLRv1 glyph given
+    /// the current transformation matrix (as accumulated by
+    /// [`push_transform`](ColorPainter::push_transform) calls).
+    fn draw_color_glyph(&mut self, _glyph: GlyphId) -> Result<ColrGlyphDrawResult, ReadError> {
+        Ok(ColrGlyphDrawResult::ColrGlyphDrawingNotImpl)
+    }
+
+    // TODO(drott): Add an optimized callback function combining clip, fill and transforms.
+
+    /// Open a new layer, and merge the layer down using `composite_mode` when
+    /// [`pop_layer`](ColorPainter::pop_layer) is called, signalling that this layer is done drawing.
+    fn push_layer(&mut self, composite_mode: CompositeMode);
+    fn pop_layer(&mut self);
+}
+
+/// Distinguishes available color glyph types.
+pub enum ColorPaintableType {
+    ColrV0,
+    ColrV1,
+}
+
+/// A representation of a color glyph that can be painted through a sequence of [`ColorPainter`] callbacks.
+#[derive(Clone)]
+pub struct ColorPaintable<'a> {
+    colr: colr::Colr<'a>,
+    root_paint_ref: ColorPaintableRoot<'a>,
+}
+
+#[derive(Clone)]
+enum ColorPaintableRoot<'a> {
+    V0Range(Range<usize>),
+    V1Paint(colr::Paint<'a>, PaintId, GlyphId, Result<u16, ReadError>),
+}
+
+impl<'a> ColorPaintable<'a> {
+    /// Returns the version of the color table from which this outline was
+    /// selected.
+    pub fn paintable_type(&self) -> ColorPaintableType {
+        match &self.root_paint_ref {
+            ColorPaintableRoot::V0Range(_) => ColorPaintableType::ColrV0,
+            ColorPaintableRoot::V1Paint(..) => ColorPaintableType::ColrV1,
+        }
+    }
+
+    /// Returns the bounding box. For COLRv1 paintables, this is clipbox of the
+    /// specified COLRv1 glyph, or `None` if there is
+    /// none for the particular glyph.  The `size` argument can optionally be used
+    /// to scale the bounding box to a particular font size. `location` allows
+    /// specifycing a variation instance.
+    pub fn get_bounding_box(
+        &self,
+        location: impl Into<LocationRef<'a>>,
+        size: Option<f32>,
```

Comment:
Changed to `Size` - but my criticism on `Size` accepting a 0 and using that to mean unscaled still stands. IMO Size should have a special constructor for expressing Unscaled and a size of 0 should just be one that makes everything be zero. Otherwise we have to special case incoming client size 0 sizes in various places, like I am doing now in this method. Having a client side size 0 become unscaled accidentally is surprising behavior.

Reply:
Ack, yes. I will submit a fix for Size this week.